server/src/test/java/org/elasticsearch/index/mapper/CalculatedFieldTests.java

```diff
@@ -0,0 +1,306 @@
+        DocumentMapper mapper = createDocumentMapper(topMapping(b -> {
+            b.startObject("runtime");
+            b.startObject("runtime-field").field("type", "long").endObject();
+            b.endObject();
+            b.startObject("properties");
+            b.startObject("index-field").field("type", "long").field("script", "refer-to-runtime").endObject();
+            b.endObject();
+        }));
+
+        Exception e = expectThrows(MapperParsingException.class, () -> mapper.parse(source(b -> {})));
+        assertEquals("No field found for [runtime-field] in mapping", e.getCause().getMessage());
+    }
+
+    private static Consumer<TestLongFieldScript> getLongScript(String name) {
+        if ("refer-to-runtime".equals(name)) {
+            return s -> {
+                s.emitValue((long) s.getDoc().get("runtime-field").get(0));
+            };
+        }
+        if (name.endsWith("_length")) {
+            String field = name.substring(0, name.lastIndexOf("_length"));
+            return s -> {
+                for (Object v : s.extractValuesFromSource(field)) {
+                    s.emitValue(Objects.toString(v).length());
+                }
+            };
+        }
+        if (name.endsWith("_plus_two")) {
+            String field = name.substring(0, name.lastIndexOf("_plus_two"));
+            return s -> {
+                long input = (long) s.getDoc().get(field).get(0);
+                s.emitValue(input + 2);
+            };
+        }
+        throw new UnsupportedOperationException("Unknown script [" + name + "]");
+    }
+
+    private static Consumer<TestDoubleFieldScript> getDoubleScript(String name) {
+        if (name.endsWith("_plus_two")) {
+            String field = name.substring(0, name.lastIndexOf("_plus_two"));
+            return s -> {
+                Number input = (Number) s.getDoc().get(field).get(0);
+                s.emitValue(input.doubleValue() + 2);
+            };
+        }
+        throw new UnsupportedOperationException("Unknown script [" + name + "]");
+    }
+
+    private static class TestLongFieldScript extends LongFieldScript {
+
+        final Consumer<TestLongFieldScript> executor;
+
+        TestLongFieldScript(
+            String fieldName,
+            Map<String, Object> params,
+            SearchLookup searchLookup,
+            LeafReaderContext ctx,
+            Consumer<TestLongFieldScript> executor
+        ) {
+            super(fieldName, params, searchLookup, ctx);
+            this.executor = executor;
+        }
+
+        @Override
+        public void execute() {
+            executor.accept(this);
+        }
+
+        public void emitValue(long v) {
+            super.emit(v);
+        }
+
+        public List<Object> extractValuesFromSource(String path) {
+            return super.extractFromSource(path);
+        }
+    }
+
+    private static class TestDoubleFieldScript extends DoubleFieldScript {
```

Comment:
These classes make it possible to separate out the `getLongScript` and `getDoubleScript` methods above - they can't take `LongFieldScript` etc directly because `emit` and `extractFromSource` are protected methods.  Another option would be to just make them public?